corehq/apps/users/login_record.py

```diff
@@ -0,0 +1,83 @@
+from datetime import datetime
+from django_redis import get_redis_connection
+
+ONE_DAY = 60 * 60 * 24
+EPOCH = datetime.utcfromtimestamp(0)
+
+
+class LoginRecord:
+    """A thread-safe class to record login failures."""
+
+    FIELD_FAILURES = 'failures'
+    FIELD_TS = 'ts'
+
+    @classmethod
+    def get(cls, username):
+        raw_failures, raw_ts = cls._get_raw_fields(username)
+        attempts_ts = float(raw_ts or 0)
+        num_failures = int(raw_failures or 0)
+
+        last_attempt_date = datetime.utcfromtimestamp(attempts_ts or 0)
+
+        return LoginRecord(username, num_failures, last_attempt_date)
+
+    @classmethod
+    def _get_raw_fields(cls, username):
+        redis_client = get_redis_connection()
+        return redis_client.hmget(cls._get_key(username), cls.FIELD_FAILURES, cls.FIELD_TS)
+
+    @staticmethod
+    def _get_key(username):
+        return f'attempts_{username}'
+
+    def __init__(self, username, failures=0, last_attempt_date=None):
+        self.username = username
+        self._reset_values(failures, last_attempt_date)
+
+        self.key = self._get_key(self.username)
+
+    def _reset_values(self, failures=0, last_attempt_date=None):
+        self.failures = failures
+        self.last_attempt_date = last_attempt_date or EPOCH
+
+    def clear(self):
+        redis_client = get_redis_connection()
+        redis_client.delete(self.key)
+
+        self._reset_values()
+
+    def add_failure(self, current_time):
+        redis_client = get_redis_connection()
+        transaction_results = redis_client.transaction(
+            self._create_failure_handler(current_time),
+            self.key)
+        num_failures = transaction_results[0]
+
+        if num_failures is True:
```

Comment:
What's the expected return type saved in `transaction_results`? `(int,)` if there are previous failed attempts and `(True,)` if there aren't? I don't quite follow where the `True` comes from.

Reply:
Yeah, this is the one real sore spot of the PR -- I couldn't think of an obvious way to convey it. `EXECUTE` (and thus transaction) return an array of all the operations performed, in order. The first thing we do is update the failures -- if the key doesn't yet exist, then we perform a `SET`, which returns True. If it already exists, then we perform `INCR`, which returns the value it incremented to. I agree this isn't very intuitive, so if anyone can think of a better way, I'm all ears

Reply:
I haven't thought this through all the way, but according to https://redis.io/commands/incr, INCR doesn't fail if the key doesn't yet have a value; it sets it to 0 first and then does the increment. So I think if we just rely on that behavior, and skip the SET below, it'll do exactly this in a less confusing way? It's very intuitive default behavior so I feel okay relying on it, as long as I'm not missing anything about how it affects the logic

Reply:
Hmmm, I misspoke. You're right -- if the field doesn't yet exist, `INCR` will handle it. However, `SET` is necessary if the previous login failures have expired (they were from yesterday). If we have yesterday's information, the number of failures needs to either be set to 1, or set to 0 and then incremented to 1. I'm handling it in the latter way for now. I could imagine implementing it in the former way, and then searching the results for the first integer, but I don't know that that is any more intuitive. It would also be very slightly slower, since we'd need redis to perform a delete and then an `INCR`